### PR TITLE
🐛  Set cdc record subsequent record wait time to initial wait time as a workaround

### LINKED ISSUE #35059: source mongodb wrapping up syncs due to heartbeats not behaving as expected
### Connector Name

source-mongodb-v2

### Connector Version

1.2.7

### What step the error happened?

During the sync

### Relevant information

On a large and rapidly updated oplog, out of which we're only reading a number of collections.
Attempting to get CDC changes from a resume token is behaving as follows:
1. We receive exactly 101 events from debezium.
2. A single or maximum two heartbeat events received.
3. The debezium connector goes into a waiting period.
4. After a while events start streaming.
No other heartbeats received

If the waiting period happens to be greater than 60 seconds, source-mogodb will start a shutdown as the `subsequentWaitTime` is hard coded as 60 seconds.
This is seen on this [workspace](https://cloud.airbyte.com/workspaces/fcf1bac7-6c6c-4a58-aa89-e73a381ea089/connections/aaff516c-cf3b-45ec-8c84-0af838bbad73/job-history), as well as with another OSS customer.

Since we never increment our resume token by more than 101 events, we eventually fall far behind in the log and have to start over with a full read of the collection.

Waiting longer than 60 seconds solves the issue.

The debezium heartbeat we configure debezium to send every 10 seconds was the signal for us to keep waiting for events. Since they aren't received it creates this problem

### Relevant log output

_No response_

### Contribute

- [ ] Yes, I want to contribute

### PR DESCRIPTION
**Pull request by rodireich**
## What
Since we can't rely on debezium heartbeats to helps us know when to wrap up our sync.
Let's make the subsequent wait time configurable.
## How
Set subsequent wait time for records to 1/2 of configured initials time.

### CODE REVIEW

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java [hunk 1]
```diff
@@ -80,8 +79,14 @@ public List<AutoCloseableIterator<AirbyteMessage>> createCdcIterators(
                                                                         final Instant emittedAt,
                                                                         final MongoDbSourceConfig config) {
 
-    final Duration firstRecordWaitTime = RecordWaitTimeUtil.getFirstRecordWaitTime(config.rawConfig());
-    final Duration subsequentRecordWaitTime = RecordWaitTimeUtil.getSubsequentRecordWaitTime(config.rawConfig());
+//    LOGGER.info("*** config {}", config.rawConfig());
```
Comment (akashkulk) [inline airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java:82]: nit : remove this commented out log line

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/MongoConstants.java [hunk 2]
```diff
@@ -34,6 +34,8 @@ public class MongoConstants {
   public static final String USERNAME_CONFIGURATION_KEY = MongoDbDebeziumConstants.Configuration.USERNAME_CONFIGURATION_KEY;
   public static final String SCHEMA_ENFORCED_CONFIGURATION_KEY = MongoDbDebeziumConstants.Configuration.SCHEMA_ENFORCED_CONFIGURATION_KEY;
   public static final String SCHEMALESS_MODE_DATA_FIELD = Configuration.SCHEMALESS_MODE_DATA_FIELD;
+  public static final String INITIAL_RECORD_WAITING_TIME_SEC = "initial_waiting_seconds";
```
Comment (akashkulk) [inline airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/MongoConstants.java:37]: do we not define initial waiting time in the spec already?

Comment (akashkulk) [reply airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/MongoConstants.java:37]: image

Comment (akashkulk) [reply airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/MongoConstants.java:37]: For mongo cdc, I think we'll always be assured that initial wait time in seconds is populated, so we can remove the defaults

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java [hunk 3]
```diff
@@ -80,8 +79,14 @@ public List<AutoCloseableIterator<AirbyteMessage>> createCdcIterators(
...
+    final Duration firstRecordWaitTime = Duration.ofSeconds(config.getInitialWaitingTimeSeconds());
+    // #35059: debezium heartbeats are not sent on the expected interval. this is a worksaround to allow
+    // making
+    // subsequent wait time configurable.
+    // final Duration subsequentRecordWaitTime = firstRecordWaitTime.dividedBy(2);
```
Comment (akashkulk) [inline airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java:87]: Is the goal to make the subsequent record wait time half of the first record wait time or the same?

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java [hunk 4]
```diff
@@ -80,8 +79,14 @@ public List<AutoCloseableIterator<AirbyteMessage>> createCdcIterators(
...
+    final Duration subsequentRecordWaitTime = firstRecordWaitTime;
+    LOGGER.info("*** subs {} init {}", subsequentRecordWaitTime, firstRecordWaitTime);
```
Comment (akashkulk) [inline airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java:89]: Can we have a log line here like : "Subsequent record wait time is : {}" to make it more clear?

**Review by akashkulk — CHANGES_REQUESTED** [review]
Another higher level question I had - do we want to make this change across all our connectors? Doesn't have to be in this PR, but wanted to get your thoughts on that

**Comment by rodireich** [comment]
> Another higher level question I had - do we want to make this change across all our connectors? Doesn't have to be in this PR, but wanted to get your thoughts on that

Heartbeats makes other connectors not rely on the subsequent wait timeout as getting scheduled hearbeats showing the progress made on the respective CDC log is letting us know when to wrap up.
There's an open thread with debezium to make mongodb heartbeat work the same (and also match their documentation).
I see this change as more of a stopgap until we have reliable debezium heartbeats.

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/MongoConstants.java (hunk 2 above)
Comment (rodireich) [reply airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/MongoConstants.java:37]: > do we not define initial waiting time in the spec already?
I'm not sure I get it
Value is defined in spec but to read it I needed to define a constant in java.
The other place reading it - incorrectly for mongodb - is `RecordWaitTimeUtil.getFirstRecordWaitSeconds()`
but no constant defined there
```
final int seconds = config.get("replication_method").get("initial_waiting_seconds").asInt();
```

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java [hunk 5]
```diff
@@ -80,8 +79,11 @@ public List<AutoCloseableIterator<AirbyteMessage>> createCdcIterators(
                                                                         final Instant emittedAt,
                                                                         final MongoDbSourceConfig config) {
 
-    final Duration firstRecordWaitTime = RecordWaitTimeUtil.getFirstRecordWaitTime(config.rawConfig());
-    final Duration subsequentRecordWaitTime = RecordWaitTimeUtil.getSubsequentRecordWaitTime(config.rawConfig());
+    final Duration firstRecordWaitTime = Duration.ofSeconds(config.getInitialWaitingTimeSeconds());
+    // #35059: debezium heartbeats are not sent on the expected interval. this is
+    // a worksaround to allow making subsequent wait time configurable.
+    final Duration subsequentRecordWaitTime = firstRecordWaitTime;
```
Comment (rodireich) [inline airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/cdc/MongoDbCdcInitializer.java:85]: a max of 10 minutes may not be enough.
Customer is still seeing 101 records and timeout with 10 minutes.

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/MongoConstants.java (hunk 2 above)
Comment (akashkulk) [reply airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/MongoConstants.java:37]: Ack - that is unfortunate

File: airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/MongoDbSourceConfig.java [hunk 6]
```diff
@@ -27,33 +29,32 @@
...
+    if (rawConfig == null) {
+      throw new IllegalArgumentException("MongoDbSourceConfig cannot accept a null config.");
+    }
+    if (!rawConfig.hasNonNull(DATABASE_CONFIG_CONFIGURATION_KEY)) {
       throw new IllegalArgumentException("Database configuration is missing required '" + DATABASE_CONFIG_CONFIGURATION_KEY + "' property.");
     }
   }
 
+  public JsonNode getDatabaseConfig() {
+    return rawConfig.get(DATABASE_CONFIG_CONFIGURATION_KEY);
```
Comment (akashkulk) [inline airbyte-integrations/connectors/source-mongodb-v2/src/main/java/io/airbyte/integrations/source/mongodb/MongoDbSourceConfig.java:42]: this seems to be the same as what it was before : 
`this.rawConfig = rawConfig.get(DATABASE_CONFIG_CONFIGURATION_KEY);`

**Review by akashkulk — APPROVED** [review]

**Pull request merged**